services/core-web/src/components/noticeOfWork/applications/FinalPermitDocuments.js

```diff
@@ -0,0 +1,235 @@
+import { COLOR } from "@/constants/styles";
+import CustomPropTypes from "@/customPropTypes";
+import NOWDocuments from "@/components/noticeOfWork/applications//NOWDocuments";
+
+/**
+ * @class FinalPermitDocuments- call logic surrounding adding or removing documents in the final Permit document list
+ */
+
+const propTypes = {
+  mineGuid: PropTypes.string.isRequired,
+  noticeOfWork: CustomPropTypes.importedNOWApplication.isRequired,
+  updateNoticeOfWorkApplication: PropTypes.func.isRequired,
+  fetchImportedNoticeOfWorkApplication: PropTypes.func.isRequired,
+  closeModal: PropTypes.func.isRequired,
+  openModal: PropTypes.func.isRequired,
+  setNoticeOfWorkApplicationDocumentDownloadState: PropTypes.func.isRequired,
+  documentDownloadState: CustomPropTypes.documentDownloadState.isRequired,
+};
+
+export class FinalPermitDocuments extends Component {
+  state = {
+    cancelDownload: false,
+  };
+
+  createFinalDocumentPackage = (selectedCoreRows) => {
+    const documentPayload = this.props.noticeOfWork.documents.map((document) => {
+      // eslint-disable-next-line no-unused-expressions
+      selectedCoreRows.includes(document.now_application_document_xref_guid)
+        ? (document.is_final_package = true)
+        : (document.is_final_package = false);
+      return document;
+    });
+    const payload = { ...this.props.noticeOfWork, documents: documentPayload };
+    const message = "Successfully updated the final application package.";
+
+    this.props
+      .updateNoticeOfWorkApplication(payload, this.props.noticeOfWork.now_application_guid, message)
+      .then(() => {
+        this.props
+          .fetchImportedNoticeOfWorkApplication(this.props.noticeOfWork.now_application_guid)
+          .then(() => {
+            this.props.closeModal();
+          });
+      });
+  };
+
+  cancelDownload = () => {
+    this.setState({ cancelDownload: true });
+  };
+
+  downloadDocument = (url) => {
+    const a = document.createElement("a");
+    a.href = url.url;
+    a.download = url.filename;
+    a.style.display = "none";
+    document.body.append(a);
+    a.click();
+    a.remove();
+  };
+
+  waitFor = (conditionFunction) => {
+    const poll = (resolve) => {
+      if (conditionFunction()) resolve();
+      else setTimeout(() => poll(resolve), 400);
+    };
+
+    return new Promise(poll);
+  };
+
+  downloadDocumentPackage = () => {
+    const docURLS = [];
+    const permitDocuments = this.props.noticeOfWork.documents.filter(
+      ({ is_final_package }) => is_final_package
+    );
+    let currentFile = 0;
+    const totalFiles = permitDocuments.length;
+    if (totalFiles === 0) return;
+
+    permitDocuments.forEach((doc) =>
+      getDocumentDownloadToken(
+        doc.mine_document.document_manager_guid,
+        doc.mine_document.document_name,
+        docURLS
+      )
+    );
+
+    this.waitFor(() => docURLS.length === permitDocuments.length).then(async () => {
```

Comment:
So we trigger the download of a file every 2 seconds. Once all file downloads have been triggered, we consider the package to have been downloaded? Just want to make sure I'm understanding the logic.

Reply:
It would be nice if there was a way to capture the callback of the user hitting "Save", because even if the user hasn't clicked "Save" on the first document, the progress bar continues on every 2 seconds. 🤔

Reply:
I copied from the download 'referral/consultation" step, 

I believe, and @JustinMacaulay can correct me if I'm wrong, Originally this check made sure that the total length of documents being downloaded matched up from the various sections (ie submission documents, and core documents) 

Now that im looking at it, I don't think this check is needed any longer. 

@JustinMacaulay do you mind taking a look at this function?